dot/peerset/peerstate.go

```diff
@@ -258,61 +289,73 @@ func (ps *PeersState) hasFreeIncomingSlot(set int) bool {
 // addNoSlotNode adds a node to the list of nodes that don't occupy slots.
 // has no effect if the node was already in the group.
 func (ps *PeersState) addNoSlotNode(idx int, peerID peer.ID) error {
+	ps.Lock()
+	defer ps.Unlock()
+
 	if _, ok := ps.sets[idx].noSlotNodes[peerID]; ok {
 		logger.Debugf("peer %s already exists in no slot node", peerID)
 		return nil
 	}
 
 	// Insert peerStatus
 	ps.sets[idx].noSlotNodes[peerID] = struct{}{}
-	n, err := ps.getNode(peerID)
-	if err != nil {
-		return fmt.Errorf("could not get node for peer id %s: %w", peerID, err)
+
+	node, has := ps.nodes[peerID]
+	if !has {
+		return fmt.Errorf("could not get node for peer id %s: %w", peerID, ErrPeerDoesNotExist)
 	}
 
-	switch n.state[idx] {
+	switch node.state[idx] {
 	case ingoing:
 		ps.sets[idx].numIn--
 	case outgoing:
 		ps.sets[idx].numOut--
 	}
 
-	ps.nodes[peerID] = n
 	return nil
 }
 
 func (ps *PeersState) removeNoSlotNode(idx int, peerID peer.ID) error {
+	ps.Lock()
+	defer ps.Unlock()
+
 	if _, ok := ps.sets[idx].noSlotNodes[peerID]; !ok {
 		logger.Debugf("peer %s is not in no-slot node map", peerID)
 		return nil
 	}
 
 	delete(ps.sets[idx].noSlotNodes, peerID)
-	n, err := ps.getNode(peerID)
-	if err != nil {
-		return fmt.Errorf("could not get node for peer id %s: %w", peerID, err)
+
+	node, has := ps.nodes[peerID]
+	if !has {
+		return fmt.Errorf("could not get node for peer id %s: %w", peerID, ErrPeerDoesNotExist)
```

Comment:
a bit of a SuPeRNiT...

`could not get node for peer id` should in the sentinel error `ErrPeerDoesNotExist`, which may be renamed to `ErrNodeNotFoundForPeer` to have i.e.:

```suggestion
		return fmt.Errorf("%w: for peer id %s", ErrNodeNotFoundForPeer, peerID)
```

wrap an error as the rightmost part, but wrap a first error as the leftmost part.